### PR TITLE
Add dask.array.random.RandomState

### LINKED ISSUE #311: Consistency in da.random
Currently multiple calls to `da.random` functions don't produce consistent values

```python
In [1]: import dask.array as da

In [2]: x = da.random.random((2,), chunks=2)

In [3]: x.compute()
Out[3]: array([ 0.47001845,  0.54208932])

In [4]: x.compute()
Out[4]: array([ 0.23277611,  0.312902  ])
```

We should correct this, either by using `np.random.seed` in each block (perhaps with a lock around the pair of calls?) or with the RandomState API.

cc @shoyer @ogrisel

### PR DESCRIPTION
**Pull request by mrocklin**
Fixes #311 

This adds a `RandomState` object to `da.random` to provide deterministic pseudorandom number generation both within each computation and between different calls to random functions.

cc @ogrisel

Example
------

```python
In [1]: import dask.array as da

In [2]: x = da.random.RandomState(42).normal(10, 0.1, size=10, chunks=5)

In [3]: x.compute()
Out[3]: 
array([ 10.12508067,   9.98296609,   9.97988253,   9.95198302,
         9.8102533 ,  10.15719861,   9.93794495,   9.98066472,
        10.04543759,   9.91149584])

In [4]: x = da.random.RandomState(42).normal(10, 0.1, size=10, chunks=5)

In [5]: x.compute()
Out[5]: 
array([ 10.12508067,   9.98296609,   9.97988253,   9.95198302,
         9.8102533 ,  10.15719861,   9.93794495,   9.98066472,
        10.04543759,   9.91149584])

In [6]: x.dask  # Note the different seeds to each numpy random call
Out[6]: 
{('x_1', 0): (<function dask.array.random.apply_random>,
  <method 'normal' of 'mtrand.RandomState' objects>,
  1608637542,
  (5,),
  (10, 0.1),
  {}),
 ('x_1', 1): (<function dask.array.random.apply_random>,
  <method 'normal' of 'mtrand.RandomState' objects>,
  1273642419,
  (5,),
  (10, 0.1),
  {})}
```

### CODE REVIEW

**Comment by mrocklin** [comment]
So far this only implements a single operation.  I'll do the necessary copy-pasting or metaprogramming once we've verified that this is the right interface.

File: dask/array/random.py [hunk 1]
```diff
@@ -54,3 +54,53 @@
...
+        if not isinstance(size, (tuple, list)):
+            size = (size,)
+
+        chunks = kwargs.pop('chunks', None)
+        chunks = normalize_chunks(chunks, size)
+        name = kwargs.pop('name', next(names))
+
+        dtype = kwargs.pop('dtype', None)
+        if dtype is None:
+            kw = kwargs.copy(); kw['size'] = (0,)
```
Comment (shoyer) [inline dask/array/random.py]: A semicolon, really? :)

**Comment by shoyer** [comment]
This looks like the right API. It's possibly worth some smoke tests to verify your parsing logic matches that of NumPy's (which is a mess, unfortunately).

I assume that one of the next steps is making `dask.array.random.uniform` use a global `RandomState` object under the hood? It's also worth checking that this is controllable with `numpy.random.seed`.

**Comment by mrocklin** [comment]
I've taken the copy-paste approach rather than metaprogramming.  This should handle any concerns about API parsing logic.

File: dask/array/tests/test_random.py [hunk 2]
```diff
@@ -39,16 +55,6 @@ def test_kwargs():
...
 
-def test_kwargs_size_or_shape():
-    a = normal(loc=10.0, scale=0.1, shape=(10, 10), chunks=(5, 5))
-    b = normal(loc=10.0, scale=0.1, size=(10, 10), chunks=(5, 5))
-    assert isinstance(a, Array)
-    assert isinstance(b, Array)
-    assert a.chunks == b.chunks
-
-    assert np.array(a).shape == np.array(b).shape
-
```
Comment (mrocklin) [inline dask/array/tests/test_random.py]: Note that this is a slight API break.

**Comment by mrocklin** [comment]
@shoyer this could use your review.  There is a fair amount of repetition here but it's straightforward/non-magical.

File: dask/array/tests/test_random.py (hunk 2 above)
Comment (ogrisel) [inline dask/array/tests/test_random.py]: why is this broken?

Comment (mrocklin) [inline dask/array/tests/test_random.py]: Previously `dask.array.random` functions supported either `size=` or `shape=` kwargs for all functions.  Now it supports only `size=` as is standard within `np.random`.

Comment (ogrisel) [inline dask/array/tests/test_random.py]: ah indeed, I missed it. I think being consistent with numpy API is good. However I also think that numpy should deprecated `size=` in favor of `shape=` :).

File: dask/array/random.py [hunk 3]
```diff
@@ -1,56 +1,284 @@
...
+
+        # Get dtype
+        kw = kwargs.copy()
+        kw['size'] = (0,)
+        dtype = func(np.random.RandomState(), *args, **kw).dtype
+
+        # Build graph
+        keys = product([name], *[range(len(bd)) for bd in chunks])
+        sizes = product(*chunks)
+        vals = ((apply_random, func, self.state.randint(2**31),
```
Comment (ogrisel) [inline dask/array/random.py]: I think you can use `2 ** 32 - 1` or even better: `np.iinfo(np.uint32).max` here.

File: dask/array/random.py [hunk 4]
```diff
@@ -1,56 +1,284 @@
...
-wald = wrap(np.random.wald)
-weibull = wrap(np.random.weibull)
-zipf = wrap(np.random.zipf)
 
-"""
-Standard distributions
-"""
+class RandomState(object):
+    def __init__(self, seed=None):
+        self.state = np.random.RandomState(seed)
```
Comment (ogrisel) [inline dask/array/random.py]: Maybe this should be a private attribute `_state` or even `_numpy_state`.

File: dask/array/random.py (hunk 3 above)
Comment (mrocklin) [inline dask/array/random.py]: Fixed

File: dask/array/random.py (hunk 4 above)
Comment (mrocklin) [inline dask/array/random.py]: Fixed

**Comment by ogrisel** [comment]
It is my understanding that this implementation is individually and deterministically seeding the dsk tasks so as to ensure reproducible results even with non-deterministic concurrent scheduler. This is great. It would be great to add a test that checks that property. E.g. based on the following untested snippet:

```python

samples_1 = da.random.RandomState(42).normal(size=1000, chunks=10)
samples_2 = da.random.RandomState(42).normal(size=1000, chunks=10)

assert samples_1.dask = samples_2.dask

# check that we get deterministic samples even for non-deterministic concurrent
# schedulers
for get, num_workers in product([dask.threaded.get, dask.multiprocessing.get], range(4)):
    assert_array_almost_equals(samples_1.compute(get=get),
                                               samples_2.compute(get=get))
```

**Comment by mrocklin** [comment]
Nice test.  I've added the dask graph checking (though I've made it blind to keys, which have auto-incrementing names for the moment.

This uncovered an issue with dill https://github.com/uqfoundation/dill/issues/109 . 

We can get around this short term by using strings rather than RandomState methods I suppose.

**Comment by ogrisel** [comment]
I think it would be good to add the test that checks that the generated values are actually deterministic on a concurrent backend such as multiprocessing. This test can be skipped under Python 2 as long as there is no fix or workaround for uqfoundation/dill#109 .

**Comment by mrocklin** [comment]
@ogrisel see recent commits

**Comment by ogrisel** [comment]
LGTM!

File: dask/array/random.py [hunk 5]
```diff
@@ -1,56 +1,314 @@
...
+        return self.wrap(np.random.RandomState.weibull, a,
+                         size=size, chunks=chunks)
+
+    @doc_wraps(np.random.RandomState.zipf)
+    def zipf(self, a, size=None, chunks=None):
+        return self.wrap(np.random.RandomState.zipf, a,
+                         size=size, chunks=chunks)
+
+
+def apply_random(func, seed, size, args, kwargs):
```
Comment (shoyer) [inline dask/array/random.py]: I would mark this as private API

File: dask/array/random.py [hunk 6]
```diff
@@ -1,56 +1,314 @@
...
+    >>> x.compute()
+    array([  9.95487579,  10.02999135,  10.08498441])
+
+    See Also:
+        np.random.RandomState
+    """
+    def __init__(self, seed=None):
+        self._numpy_state = np.random.RandomState(seed)
+
+    def wrap(self, func, *args, **kwargs):
```
Comment (shoyer) [inline dask/array/random.py]: I would also probably make this private

File: dask/array/random.py (hunk 5 above)
Comment (mrocklin) [inline dask/array/random.py]: Fixed

File: dask/array/random.py (hunk 6 above)
Comment (mrocklin) [inline dask/array/random.py]: made both private

**Comment by mrocklin** [comment]
This is in.  Thanks all for the review.

**Comment by cpcloud** [comment]
I wonder if this may have fixed the `random_split` issue.

**Comment by cpcloud** [comment]
Nope

**Pull request merged**